cmd/oras/root/pull.go

```diff
@@ -105,30 +106,77 @@ Example - Pull artifact files from an OCI layout archive 'layout.tar':
 func runPull(ctx context.Context, opts pullOptions) error {
 	ctx, logger := opts.WithContext(ctx)
 	// Copy Options
-	var printed sync.Map
 	copyOptions := oras.DefaultCopyOptions
 	copyOptions.Concurrency = opts.concurrency
+	if opts.Platform.Platform != nil {
+		copyOptions.WithTargetPlatform(opts.Platform.Platform)
+	}
+	target, err := opts.NewReadonlyTarget(ctx, opts.Common, logger)
+	if err != nil {
+		return err
+	}
+	if err := opts.EnsureReferenceNotEmpty(); err != nil {
+		return err
+	}
+	src, err := opts.CachedTarget(target)
+	if err != nil {
+		return err
+	}
+	dst, err := file.New(opts.Output)
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+	dst.AllowPathTraversalOnWrite = opts.PathTraversal
+	dst.DisableOverwrite = opts.KeepOldFiles
+
+	desc, pulledEmpty, err := opts.doPull(ctx, src, dst, copyOptions)
+	if err != nil {
+		if errors.Is(err, file.ErrPathTraversalDisallowed) {
+			err = fmt.Errorf("%s: %w", "use flag --allow-path-traversal to allow insecurely pulling files outside of working directory", err)
+		}
+		return err
+	}
+	if pulledEmpty {
+		fmt.Println("Downloaded empty artifact")
+	}
+	fmt.Println("Pulled", opts.AnnotatedReference())
+	fmt.Println("Digest:", desc.Digest)
+	return nil
+}
+
+func (po *pullOptions) doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget, opts oras.CopyOptions) (ocispec.Descriptor, bool, error) {
 	var configPath, configMediaType string
 	var err error
-	if opts.ManifestConfigRef != "" {
-		configPath, configMediaType, err = fileref.Parse(opts.ManifestConfigRef, "")
+	if po.ManifestConfigRef != "" {
+		configPath, configMediaType, err = fileref.Parse(po.ManifestConfigRef, "")
 		if err != nil {
-			return err
+			return ocispec.Descriptor{}, false, err
 		}
 	}
-	if opts.Platform.Platform != nil {
-		copyOptions.WithTargetPlatform(opts.Platform.Platform)
+
+	var tracked track.GraphTarget
+	if po.TTY != nil {
+		tracked, err = track.NewTarget(dst, "Downloading", "Downloaded ", po.TTY)
+		if err != nil {
+			return ocispec.Descriptor{}, false, err
+		}
+		defer tracked.Close()
+		dst = tracked
 	}
```

Comment:
nit: Like `getTrackedTarget`, it seems can also be refactored although it is invoked only once.